### ISSUE TITLE
HttpClient - Kerberos authentication fails after a redirect

### ISSUE THREAD

**Issue by manuelspezzani** [dotnet/runtime#89300]
### Description

I'm using an HttpClient to call an internal REST API secured with Kerberos authentication.
The API is behind a network load balancer that performs http to https redirection.

Https redirection happens as expected, but the client doesn't authenticate even if the server sends a valids authentication challenge.
Using the https endpoint directly (skipping the redirection step) works as expected.

### Reproduction Steps

```
internal class Program
{
	static async Task Main(string[] args)
	{
		var client = new HttpClient(new HttpClientHandler{UseDefaultCredentials = true});
		var response = await client.GetAsync("http://myservice.company.com/v1/whatever");

		response.EnsureSuccessStatusCode(); // this fails since the status code is 401
	}
}
```

### Expected behavior

The client should follow the redirection, then it should authenticate using the provided network credentials.

### Actual behavior

Calling the http endpoint produces the following:
```
> GET http://myservice.company.com/v1/whatever HTTP/1.1
> Host: myservice.company.com

< HTTP/1.1 307 Temporary Redirect
< Location: https://myservice.company.com/v1/whatever HTTP/1.1
< Connection: Keep-Alive
< Content-Length: 0

> GET https://myservice.company.com/v1/whatever HTTP/1.1
> Host: myservice.company.com

< HTTP/1.1 401 Unauthorized
< Content-Length: 0
< Date: Fri, 21 Jul 2023 12:21:27 GMT
< Server: Kestrel
< Set-Cookie: _40799=19a80d8c494ce8ff; Path=/
< Www-Authenticate: Negotiate
< Www-Authenticate: Bearer
< Proxy-Support: Session-Based-Authentication
```

The client doesn't "react" to that authentication challenge and returns an HTTP 401 status code response.

### Regression?

I've tested .NET Core 3.1 and .NET 6.0: both present the same behavior.
The same code works for both http and https endpoints on .NET Framework 4.8.

### Known Workarounds

The only workaround I've found is disable automatic redirection and manually implement it on a Delegating handler: 

```
internal class Program
{
	static async Task Main(string[] args)
	{
		var httpClientHandler = new HttpClientHandler
		{
			UseDefaultCredentials = true,
			AllowAutoRedirect = false
		};

		var client = new HttpClient(new ManualRedirectHandler(httpClientHandler));
		var response = await client.GetAsync("http://myservice.company.com/v1/whatever");

		response.EnsureSuccessStatusCode();
	}
}

internal class ManualRedirectHandler : DelegatingHandler
{
    public ManualRedirectHandler(HttpMessageHandler innerHandler) : base(innerHandler)
    {
    }

    public int MaxRedirections { get; set; } = 50;

    protected override HttpResponseMessage Send(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        var response = base.Send(request, cancellationToken);
        var retryCount = 0;

        while (IsRedirect(response) && retryCount < MaxRedirections)
        {
            retryCount++;
            request.RequestUri = response.Headers.Location;
            response = base.Send(request, cancellationToken);
        }

        return response;
    }

    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        var response = await base.SendAsync(request, cancellationToken);
        var retryCount = 0;

        while (IsRedirect(response) && retryCount < MaxRedirections)
        {
            retryCount++;
            request.RequestUri = response.Headers.Location;
            response = await base.SendAsync(request, cancellationToken);
        }

        return response;
    }

    private static bool IsRedirect(HttpResponseMessage response)
    {
        return response.StatusCode is HttpStatusCode.TemporaryRedirect or HttpStatusCode.PermanentRedirect or HttpStatusCode.Redirect;
    }
}
```

### Configuration

- .NET 6.0
- Windows 10 Enterprise LTSC Version 1809
- x64 architecture

I didn't tried on other platformt: I'm not sure if it's specific to this configuration

### Other information

Not sure if it is related but, if you look at the HTTP trace, you'll see that the authentication challenge contains two options: Negotiate and Bearer.
Maybe I'm wrong, but I don't expect it to be an issue because:
1. when https redirection is not involved, everything works as expected
2. Negotiate is the preferred option

**Comment by wfurt** [comment]
Are you also on Windows @ross-lanes? I'm sure we can figure it out ion we get repro or better insight. 
In the example above this original and redirected URL is the same - perhaps of stripped context. 
Is the host same e.g. same SPN?

**Comment by ross-lanes** [comment]
yes, windows 10 and Server 2019 Standard, I didn't consider trying Linux.   All hosts do have the same SPN. 

Sadly this is all part of a private repository right now, but I can share with you the request flow, I was able to use Fiddler to capture the traffic. 

so in NET8

**GET** /resource - 401
_Request Headers_:
Accept-Encoding: gzip, deflate
traceparent: 00-03e9e515736fd88ccddccb5fa33aa621-13f3aa7f97b1c291-00

_Response Headers_:
Date: Wed, 13 Mar 2024 18:19:08 GMT
Server: Apache
X-Xss-Protection: 1; mode=block
Strict-Transport-Security: max-age=15768000
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Permitted-Cross-Domain-Policies: none
WWW-Authenticate: Negotiate
WWW-Authenticate: Basic realm="PNNL Kerberos Authentication"
Content-Length: 381
Content-Type: text/html; charset=iso-8859-1
Proxy-Support: Session-Based-Authentication

**GET** /resource - 302
_Request Headers_:
Accept-Encoding: gzip, deflate
traceparent: 00-03e9e515736fd88ccddccb5fa33aa621-13f3aa7f97b1c291-00
Authorization: Negotiate <SECRET>

_Response Headers_:
Date: Wed, 13 Mar 2024 18:19:08 GMT
Server: Apache
WWW-Authenticate: Negotiate oYG2MIGzoAMKAQChCwYJKoZIgvcSAQICooGeBIGbYIGYBgkqhkiG9xIBAgICAG+BiDCBhaADAgEFoQMCAQ+ieTB3oAMCARKicARuV4+TRjERvKbgwB5VejRlrc0rIIe4Ebjh93XW3466D73QpfEPD6Ut67GvreWeACbYkmDFrSeZtGTtBU5xCMQ0yzZ2VsVZw5YzI4p/y/ciB0K4K8ecdkwKAG/oTiAvFmBIK5zEqdSn4Tnt2ZI6qGQ=, Negotiate oYG2MIGzoAMKAQChCwYJKoZIgvcSAQICooGeBIGbYIGYBgkqhkiG9xIBAgICAG+BiDCBhaADAgEFoQMCAQ+ieTB3oAMCARKicARuF5AVH1YwW8fgTe8vVwohIVJl1gB0/s+KMOtARYGqC+M/ygh6XcBfVG5S1ov++iDvGJDFkfnoD5aAp+IvKoNdzIz/PsS+/oo/GqzmBPM/EWv/DJ3Q3D69FQgGgzgHDqpYNDUZ9niyXr9qZic+b+o=
X-Powered-By: PHP/8.1.27
Cache-Control: must-revalidate, no-cache, private
X-UA-Compatible: IE=edge
Content-language: en
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
Expires: Sun, 19 Nov 1978 05:00:00 GMT
X-Generator: Drupal 9 (https://www.drupal.org)
X-Xss-Protection: 1; mode=block
Strict-Transport-Security: max-age=15768000
X-Permitted-Cross-Domain-Policies: none
Location: /sso/login/kerberos?sso_destination=/resource 
Access-Control-Allow-Headers: Content-Type, Accept, X-CSRF-Token
Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS
Access-Control-Allow-Credentials: true
Transfer-Encoding: chunked
Content-Type: text/html; charset=UTF-8

**GET** /sso/login/kerberos?sso_destination=/resource  - 401
_Request Headers_:
Accept-Encoding: gzip, deflate
traceparent: 00-03e9e515736fd88ccddccb5fa33aa621-13f3aa7f97b1c291-00

_Response Headers_:
Date: Wed, 13 Mar 2024 18:19:08 GMT
Server: Apache
X-Xss-Protection: 1; mode=block
Strict-Transport-Security: max-age=15768000
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Permitted-Cross-Domain-Policies: none
WWW-Authenticate: Negotiate
WWW-Authenticate: Basic realm="PNNL Kerberos Authentication"
Content-Length: 381
Content-Type: text/html; charset=iso-8859-1
Proxy-Support: Session-Based-Authentication

In 4.8, browser and the expectation is that the HTTP Client would send another **GET** /sso/login/kerberos?sso_destination=/resource including the Authorization header internally but it returns the 401 response back to my code. 

To work around this we have added the following (Note this is demo code based on our actual solution, in reality we are using IHttpClientFactory).

```csharp
using System;
using System.Threading.Tasks;
using System.Threading;
using System.Net.Http;
using System.Net;
using System.Linq;

public class Program
{
	static async Task Main(string[] args)
	{
		var client = new HttpClient(new HttpClientHandler { UseDefaultCredentials = true });
		await ExecuteRequest(client, new Uri("http://myservice.company.com/v1/whatever"));
	}

	private static async Task<string> ExecuteRequest(HttpClient client, Uri requestUri, CancellationToken cancellationToken = default, int retryLimit = 3)
	{
		using var request = new HttpRequestMessage(HttpMethod.Get, requestUri);
		using var response = await client.SendAsync(request, cancellationToken);
		var body = await response.Content.ReadAsStringAsync();
		// I know this is a wired but this is needed till this https://github.com/dotnet/runtime/issues/89300
		if (response.StatusCode == HttpStatusCode.Unauthorized && response.Headers.TryGetValues("WWW-Authenticate", out var wwwAuthenticateHeaderValues) && wwwAuthenticateHeaderValues.Contains("Negotiate"))
		{
			if (retryLimit <= 0)
			{
				// Log an error or handle the maximum retry limit reached scenario
				throw new Exception("Retry limit reached. Unable to authenticate.");
			}

			return await ExecuteRequest(client, request.RequestUri, cancellationToken, retryLimit - 1);
		}

		response.EnsureSuccessStatusCode();
		return body; // Changed to return the previously read 'body' variable to avoid reading the content twice.
	}
}
```

**Comment by ross-lanes** [comment]
yes, windows 10 and Server 2019 Standard, I didn't consider trying Linux.   All hosts do have the same SPN. 

Sadly this is all part of a private repository right now, but I can share with you the request flow, I was able to use Fiddler to capture the traffic. 

so in NET8

**GET** /resource - 401
_Request Headers_:
Accept-Encoding: gzip, deflate
traceparent: 00-03e9e515736fd88ccddccb5fa33aa621-13f3aa7f97b1c291-00

_Response Headers_:
Date: Wed, 13 Mar 2024 18:19:08 GMT
Server: Apache
X-Xss-Protection: 1; mode=block
Strict-Transport-Security: max-age=15768000
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Permitted-Cross-Domain-Policies: none
WWW-Authenticate: Negotiate
WWW-Authenticate: Basic realm="PNNL Kerberos Authentication"
Content-Length: 381
Content-Type: text/html; charset=iso-8859-1
Proxy-Support: Session-Based-Authentication

**GET** /resource - 302
_Request Headers_:
Accept-Encoding: gzip, deflate
traceparent: 00-03e9e515736fd88ccddccb5fa33aa621-13f3aa7f97b1c291-00
Authorization: Negotiate <SECRET>

_Response Headers_:
Date: Wed, 13 Mar 2024 18:19:08 GMT
Server: Apache
WWW-Authenticate: Negotiate <SECRET>
X-Powered-By: PHP/8.1.27
Cache-Control: must-revalidate, no-cache, private
X-UA-Compatible: IE=edge
Content-language: en
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
Expires: Sun, 19 Nov 1978 05:00:00 GMT
X-Generator: Drupal 9 (https://www.drupal.org)
X-Xss-Protection: 1; mode=block
Strict-Transport-Security: max-age=15768000
X-Permitted-Cross-Domain-Policies: none
Location: /sso/login/kerberos?sso_destination=/resource 
Access-Control-Allow-Headers: Content-Type, Accept, X-CSRF-Token
Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS
Access-Control-Allow-Credentials: true
Transfer-Encoding: chunked
Content-Type: text/html; charset=UTF-8

**GET** /sso/login/kerberos?sso_destination=/resource  - 401
_Request Headers_:
Accept-Encoding: gzip, deflate
traceparent: 00-03e9e515736fd88ccddccb5fa33aa621-13f3aa7f97b1c291-00

_Response Headers_:
Date: Wed, 13 Mar 2024 18:19:08 GMT
Server: Apache
X-Xss-Protection: 1; mode=block
Strict-Transport-Security: max-age=15768000
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Permitted-Cross-Domain-Policies: none
WWW-Authenticate: Negotiate
WWW-Authenticate: Basic realm="PNNL Kerberos Authentication"
Content-Length: 381
Content-Type: text/html; charset=iso-8859-1
Proxy-Support: Session-Based-Authentication

In 4.8, browser and the expectation is that the HTTP Client would send another **GET** /sso/login/kerberos?sso_destination=/resource including the Authorization header internally but it returns the 401 response back to my code. 

To work around this we have added the following (Note this is demo code based on our actual solution, in reality we are using IHttpClientFactory).

```csharp
using System;
using System.Threading.Tasks;
using System.Threading;
using System.Net.Http;
using System.Net;
using System.Linq;

public class Program
{
	static async Task Main(string[] args)
	{
		var client = new HttpClient(new HttpClientHandler { UseDefaultCredentials = true });
		await ExecuteRequest(client, new Uri("http://myservice.company.com/v1/whatever"));
	}

	private static async Task<string> ExecuteRequest(HttpClient client, Uri requestUri, CancellationToken cancellationToken = default, int retryLimit = 3)
	{
		using var request = new HttpRequestMessage(HttpMethod.Get, requestUri);
		using var response = await client.SendAsync(request, cancellationToken);
		var body = await response.Content.ReadAsStringAsync();
		// I know this is a wired but this is needed till this https://github.com/dotnet/runtime/issues/89300
		if (response.StatusCode == HttpStatusCode.Unauthorized && response.Headers.TryGetValues("WWW-Authenticate", out var wwwAuthenticateHeaderValues) && wwwAuthenticateHeaderValues.Contains("Negotiate"))
		{
			if (retryLimit <= 0)
			{
				// Log an error or handle the maximum retry limit reached scenario
				throw new Exception("Retry limit reached. Unable to authenticate.");
			}

			return await ExecuteRequest(client, request.RequestUri, cancellationToken, retryLimit - 1);
		}

		response.EnsureSuccessStatusCode();
		return body; // Changed to return the previously read 'body' variable to avoid reading the content twice.
	}
}
```

**Comment by wfurt** [comment]
I think I understand what is going on. 

https://github.com/dotnet/runtime/blob/f529d5d2323e8ad0f485b660b392f4459de24fbe/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs#L523-L534


The `UseDefaultCredentials = true` simply turns into `handler.Credentials = CredentialCache.DefaultCredentials` and it fails that check so credentials are removed. 

As far as I can tell the workaround could be something like
```
   var cache = new CredentialCache();
   cache.Add(new Uri("http://myservice.company.com/"), "Negotiate", CredentialCache.DefaultCredentials );
   handler.Credentials = cache;
```

could you give it try @ross-lanes  and @manuelspezzani  and provide feedback. 


In meantime, I'm wondering if we could/should relax that check. Since the `DefaultCredentials` is really stub I don't see problem with leaking credentials (like basic auth) - it will work only for Negotiate e.g. Kerberos & NTLM (if enabled). 


any more thoughts on this @stephentoub @GrabYourPitchforks @blowdart ????

**Comment by blowdart** [comment]
I would never recommend a work around that sends credentials over HTTP :)

@manuelspezzani is there a reason you can't hit https directly, rather than follow the redirect? I'm not exactly at peace with relaxing the check.

**Comment by manuelspezzani** [comment]
@blowdart of course I can directly hit the https endpoint, but I've opened this issue to point out what in my opinion is an important breaking change between .NET Framework and .NET. A change that, as far as I know, is completely undocumented.

Moreover, there is no easy way to "rollback" to the old behavior, unless you consider my workaround a viable solution (I don't).
I can think of at least two scenario that are badly affected by this new behavior:

**Scenario 1 - http to https** 
We have a few really old internal services, used by both .NET and .NET Framework clients.
Those services were exposed over http only (!).
To fix that, our idea was to: 
1. configure the load balancer to expose those services over https
2. configure the load balancer to automatically redirect http calls to https 
3. gradually upgrade clients to use the https endpoints directly 

Please note that this approach allow us to immediately upgrade all connections to https. But we can't: .NET clients will break because they will refuse to authenticate. 

So, as an alternative, we have to:
1. configure the load balancer to expose the service over https AND http
2. reconfigure all clients to use the https endpoint directly 
3. configure the load balancer to reject http calls 

Problem is that, until the last client is reconfigured, we will have to keep the http endpoints active.

**Scenario 2 - migration to another infrastructure** 
We are migrating a lot of our services from IIS to Docker. 
To avoid big-bang migrations, we have both instances running at the same time and, gradually, we migrate clients from the legacy infrastructure to the new one.
Having the option to migrate clients leveraging http redirections allow us to complete the migration much faster, because of course we don't have to wait for clients to be upgrated. 
It's also much safer: in case of issues we can easily rollback all (or some) clients to the old infrastructure.



I'm definitely no security expert so please correct me if I'm wrong but, as far as I know, there is no compelling reason not to authenticate a request after a redirect when you are using "proper" Kerberos implementations. Especially when the HttpClient have been explicitly configured to automatically follow redirects...


@wfurt unfortunately your suggestion is not working for me. Since the authentication is performed _after_ the redirect, I'll have to add the _https_ uri to the credential cache. But that of course defeats the whole purpose of the http redirection: if you already know the "final" uri, then simply use it. :)

**Comment by wfurt** [comment]
> @wfurt unfortunately your suggestion is not working for me. Since the authentication is performed _after_ the redirect, I'll have to add the _https_ uri to the credential cache. But that of course defeats the whole purpose of the http redirection: if you already know the "final" uri, then simply use it. :)

Could you at least try it @manuelspezzani to verify my assumptions are right?

**Comment by manuelspezzani** [comment]
@wfurt my apologies, I've forgot to mention that I've actually tried your approach... This code indeed works (please note the _https_ uri):
```
   var cache = new CredentialCache();
   cache.Add(new Uri("https://myservice.company.com/"), "Negotiate", CredentialCache.DefaultCredentials );
   handler.Credentials = cache;
```
